Source/Core/Core/ScriptEngine.cpp

```diff
@@ -0,0 +1,1010 @@
+      return 0;
+  }
+  // Get reference to Lua function.
+  Script::Context* const ctx = get_context(L);
+  lua_pushvalue(L, 2);
+  int lua_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  // Register breakpoint on PowerPC.
+  const LuaFuncHandle handle(ctx, lua_ptr, lua_ref);
+  PowerPC::script_hooks.m_hooks.emplace(addr, handle);
+
+  return 0;
+}
+
+// Lua: unhook_instruction(number address, function hook) -> ()
+// Removes a hook previously registered with hook_instruction, with the same arguments.
+static int unhook_instruction(lua_State* L)
+{
+  const std::optional<u32> addr_arg = luaL_checkaddress(L, 1);
+  luaL_checkany(L, 2);  // hook function
+  if (!addr_arg.has_value())
+    return 0;
+  if (!lua_isfunction(L, 2))
+    return 0;
+  const u32 addr = *addr_arg;
+
+  // Get pointer to Lua function.
+  const void* const lua_ptr = lua_topointer(L, 2);
+  // Search for pointer and remove reference.
+  const auto range = PowerPC::script_hooks.m_hooks.equal_range(addr);
+  for (auto it = range.first; it != range.second; it++)
+  {
+    if (it->second.lua_ptr() == lua_ptr)
+    {
+      luaL_unref(L, LUA_REGISTRYINDEX, it->second.lua_ref());
+      PowerPC::script_hooks.m_hooks.erase(it);
+      lua_pushboolean(L, 1);
+      return 1;
+    }
+  }
+  // Nothing found.
+  lua_pushboolean(L, 0);
+  return 1;
+}
+
+// Lua: hook_frame(function hook) -> ()
+// Registers a function to be called each frame.
+static int hook_frame(lua_State* L)
+{
+  if (!lua_isfunction(L, 1))
+    return 0;
+
+  // Get pointer to Lua function.
+  const void* const lua_ptr = lua_topointer(L, 1);
+  // Skip if function is already registered.
+  if (s_frame_hooks.count(lua_ptr) > 0)
+    return 0;
+  // Get reference to Lua function.
+  Script::Context* const ctx = get_context(L);
+  lua_pushvalue(L, 1);
+  const int lua_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+  // Register frame hook.
+  const LuaFuncHandle handle(ctx, lua_ptr, lua_ref);
+  s_frame_hooks.emplace(lua_ptr, handle);
+
+  return 0;
+}
+
+// Lua: unhook_frame(function hook) -> ()
+// Removes a hook previously registered with unhook_frame, with the same arguments.
+static int unhook_frame(lua_State* L)
+{
+  if (!lua_isfunction(L, 1))
+    return 0;
+
+  const void* lua_ptr = lua_topointer(L, 1);
+  const auto entry = s_frame_hooks.find(lua_ptr);
+  if (entry == s_frame_hooks.end())
+    return 0;
+  // Dereference function.
+  luaL_unref(L, LUA_REGISTRYINDEX, entry->second.lua_ref());
+  s_frame_hooks.erase(entry);
+
+  return 0;
+}
+
+// Lua: mem_read(number address, number len) -> (string)
+// Reads bytes from PowerPC memory beginning at the provided offset.
+static int mem_read(lua_State* L)
+{
+  const std::optional<u32> addr_arg = luaL_checkaddress(L, 1);
+  const lua_Integer len_arg = luaL_checkinteger(L, 2);
+
+  if (addr_arg)
+  {
+    size_t len = 0;
+    if (len_arg > 0)
+      len = static_cast<size_t>(len_arg);
+    const std::string cpp_str = PowerPC::HostGetString(*addr_arg, len);
```

Comment:
HostGetString() stops when it reaches a null terminator, making it unsuitable for mem_read(). It would be good for implementing str_read(), though.